posthog/hogql_queries/query_runner.py

```diff
@@ -349,53 +373,89 @@ def _limit_context_aliased_for_cache(self) -> LimitContext:
+    ) -> Optional[CR | CacheMissResponse | QueryStatus]:
+        CachedResponse: type[CR] = self.cached_response_type
+        cached_response: CR | CacheMissResponse
+        cached_response_candidate_bytes: Optional[bytes] = get_safe_cache(cache_key)
+        cached_response_candidate: Optional[dict] = (
+            OrjsonJsonSerializer({}).loads(cached_response_candidate_bytes) if cached_response_candidate_bytes else None
+        )
+        if self.is_cached_response(cached_response_candidate):
+            cached_response_candidate["is_cached"] = True
+            cached_response = CachedResponse(**cached_response_candidate)
+        elif cached_response_candidate is None:
+            cached_response = CacheMissResponse(cache_key=cache_key)
+        else:
+            # Whatever's in cache is malformed, so let's treat is as non-existent
+            cached_response = CacheMissResponse(cache_key=cache_key)
+            with push_scope() as scope:
+                scope.set_tag("cache_key", cache_key)
+                capture_exception(
+                    ValueError(f"Cached response is of unexpected type {type(cached_response)}, ignoring it")
+                )
+
+        if self.is_cached_response(cached_response_candidate):
+            if not self._is_stale(cached_response):
+                QUERY_CACHE_HIT_COUNTER.labels(team_id=self.team.pk, cache_hit="hit").inc()
+                # We have a valid result that's fresh enough, let's return it
+                return cached_response
+
+            QUERY_CACHE_HIT_COUNTER.labels(team_id=self.team.pk, cache_hit="stale").inc()
+            # We have a stale result. If we aren't allowed to calculate, let's still return it
+            # – otherwise let's proceed to calculation
+            if execution_mode == ExecutionMode.CACHE_ONLY_NEVER_CALCULATE:
+                return cached_response
+            elif execution_mode == ExecutionMode.RECENT_CACHE_CALCULATE_ASYNC_IF_STALE:
+                # We're allowed to calculate, but we'll do it asynchronously and attach the query status
+                query_status_response = self.enqueue_async_calculation(cache_key=cache_key, user=user)
+                cached_response.query_status = query_status_response.query_status
+                return cached_response
+        else:
+            QUERY_CACHE_HIT_COUNTER.labels(team_id=self.team.pk, cache_hit="miss").inc()
+            # We have no cached result. If we aren't allowed to calculate, let's return the cache miss
+            # – otherwise let's proceed to calculation
+            if execution_mode == ExecutionMode.CACHE_ONLY_NEVER_CALCULATE:
+                return cached_response
+            elif execution_mode == ExecutionMode.RECENT_CACHE_CALCULATE_ASYNC_IF_STALE:
+                # We're allowed to calculate, but we'll do it asynchronously
+                return self.enqueue_async_calculation(cache_key=cache_key, user=user)
```

Comment:
One solution in terms of typing here might be adding `query_status` as a `CacheMissResponse` field

Reply:
Good idea, this way we don't have too many things.